provider/mem/mem.go

```diff
@@ -34,10 +34,11 @@ const alertChannelLength = 200
 type Alerts struct {
 	cancel context.CancelFunc
 
+	mtx sync.RWMutex
```

Comment:
It's not strictly the same, as these two locks have different scopes, and I think I will keep this change, as this reason is not convincing to me

Reply:
further improvements can continue from here

Reply:
I think there might have been some confusion here. I think changing the mutex from `sync.Mutex` to a `sync.RWMutex` is a nice performance improvement, I was just suggesting to do it in a separate PR. There are a couple of reasons for this:

1. You can make the performance improvement by changing `sync.Mutex` to `sync.RWMutex` in both the memstore and memprovider at the same time. You can also provide some benchmarks to show how much faster it is without it being lost in this discussion which is related to another issue (fixing a race condition).
2. By keeping the fix for the race condition and performance improvements in separate PRs, it makes it so much easier to see where changes came from and why they were added when running `git blame`. Otherwise it looks like the `sync.RWMutex` is somehow related to the race condition, when it's actually just a nice performance improvement. You have to instead look at the discussion on GitHub to find all the information and understand why it was added.

Reply:
I think we've already made a few changes like this. However, I can change it back if you prefer.